hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1597,6 +1597,123 @@ private class Emit(
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) * coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) * coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementAdd = lPType.elementType match {
+          case PInt32(_) => (a: Code[_], b: Code[_]) => (coerce[Int](a) + coerce[Int](b)).asInstanceOf[Code[Any]]
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) + coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) + coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) + coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementZero = lPType.elementType match {
+          case PInt32(_) => const(0).asInstanceOf[Code[Any]]
+          case PInt64(_) => const(0L).asInstanceOf[Code[Any]]
+          case PFloat32(_) => const(0.0f).asInstanceOf[Code[Any]]
+          case PFloat64(_) => const(0.0).asInstanceOf[Code[Any]]
+        }
+
+        val eVti = typeToTypeInfo(lPType.elementType.virtualType)
+
+        val emitter = new NDArrayEmitter(mb, outputPType.nDims, unifiedShapeArray, lPType.shape.pType, lPType.elementType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val seqIdxVars = idxVars.toSeq
+            val element = coerce[Any](mb.newField("matmul_element")(eVti))
+            val kLocal = mb.newLocal[Long]
+            val k = kLocal.load()
+
+            val (lIdxVars: Array[Code[Long]], rIdxVars: Array[Code[Long]]) = (lNDims, rNDims) match {
+              case (1, 1) => (Array(k), Array(k))
+              case (1, _) =>
+                val stackDims :+ m = seqIdxVars
+
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+                (Array(k), rStackVars :+ k :+ m)
+              case (_, 1) =>
+                val stackDims :+ n = seqIdxVars
+
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                (lStackVars :+ n :+ k, Array(k))
+              case (_, _) => {
+                val stackDims :+ n :+ m = seqIdxVars
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+
+                (lStackVars :+ n :+ k, rStackVars :+ k :+  m)
+              }
+            }
+
+            val lElem = Region.loadIRIntermediate(lPType.elementType)(lPType.getElementAddress(lIdxVars, leftND, region, mb))
+            val rElem = Region.loadIRIntermediate(rPType.elementType)(rPType.getElementAddress(rIdxVars, rightND, region, mb))
```

Comment:
getElementAddress is not great. it loads the data address every time. it loads the length every time. The length will be a cache miss in lots of cases. 300 * 300 * 8 is bigger than L1 cache.

numpy is definitely calling a tiled matrix multiply algorithm, but I don't think that would yield thee orders of magnitude.